src/package.rs

```diff
@@ -57,28 +65,75 @@ impl PackageStore {
             .wrap_err("Failed to uninstall dependencies")
     }
 
-    /// Installs a package into the local file system
-    pub async fn install(package: Package) -> eyre::Result<()> {
+    /// Unpacks a package into a local directory
+    pub async fn unpack(package: &Package, path: &Path) -> eyre::Result<()> {
         let mut tar = Vec::new();
 
-        let mut gz = flate2::read::GzDecoder::new(package.tgz.reader());
+        let mut gz = flate2::read::GzDecoder::new(package.tgz.clone().reader());
 
         gz.read_to_end(&mut tar)
             .wrap_err("Failed to decompress package")?;
 
         let mut tar = tar::Archive::new(Bytes::from(tar).reader());
 
-        let pkg_dir = Path::new(Self::PROTO_DEP_PATH).join(package.name.as_package_dir());
+        let pkg_dir = path.join(package.manifest.name.as_package_dir());
+
+        fs::remove_dir_all(&pkg_dir).await.ok();
 
-        Self::uninstall(&package.name).await.ok();
         fs::create_dir_all(&pkg_dir)
             .await
             .wrap_err("Failed to install dependencies")?;
 
-        tar.unpack(pkg_dir)
-            .wrap_err(format!("Failed to unpack tar of {}", package.name))?;
+        tar.unpack(pkg_dir.clone())
+            .wrap_err(format!("Failed to unpack tar of {}", package.manifest.name))?;
 
-        tracing::info!("+ installed {}@{}", package.name, package.version);
+        tracing::debug!(
+            ":: unpacked {}@{} into {}",
+            package.manifest.name,
+            package.manifest.version,
+            pkg_dir.display()
+        );
+
+        Ok(())
+    }
+
+    /// Installs a package and all of its dependency into the local filesystem
+    pub async fn install<R: Registry>(dependency: Dependency, registry: R) -> eyre::Result<()> {
+        let package = registry.download(dependency).await?;
+
+        let dep_dir = Path::new(Self::PROTO_DEP_PATH);
+
+        Self::unpack(&package, dep_dir).await?;
+
+        tracing::info!(
+            ":: installed {}@{}",
+            package.manifest.name,
+            package.manifest.version
+        );
+
+        let Manifest { dependencies, .. } = Self::resolve(&package.manifest.name).await?;
+
+        let package_dir = &dep_dir.join(package.manifest.name.as_str());
+
+        let dependency_count = dependencies.len();
+
+        for (index, dependency) in dependencies.into_iter().enumerate() {
+            let dependency = registry.download(dependency).await?;
+
+            Self::unpack(&dependency, &package_dir).await?;
+
+            let tree_char = if index + 1 == dependency_count {
+                '┗'
+            } else {
+                '┣'
+            };
+
+            tracing::info!(
+                "   {tree_char} installed {}@{}",
+                dependency.manifest.name,
+                dependency.manifest.version
+            );
```

Comment:
If there's concurrent stuff happening the output might be intertwined. Maybe write it all at once?

Reply:
there is no concurrent stuff happening, but i agree 😊 
i tried to use tracing spans but it complains about not using them across async ops

Reply:
I think you can annotate the whole function or use a closure, see https://docs.rs/tracing/latest/tracing/struct.Span.html#in-asynchronous-code.

However this won't help here, as individual log events will be pushed to your subscriber anyway, just potentially augmented with some span metadata.
Spans might be interesting nonetheless get some timing information 🤷

Reply:
Okay :) I'll fix the logging! Thx for the hint